Fix to issue #218

desireevl opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Fix to issue #218 


### Details and comments
Removed *args from QSVC and QSVR classes

adekusar-drl reviewed (commented):
Looks good. 

Please add a unit test that covers the issue. To avoid actual printing, I think, the test should have something like `_ = str(qsvc)` and no exception should be risen.

Also a reno file with a short bug fix description is required. Take a look at https://github.com/Qiskit/qiskit-machine-learning/blob/main/CONTRIBUTING.md#release-notes for more details how to create a reno file.

woodsp-ibm commented:
Now while positional arguments is deprecated in sklearn and args should be removed, by removing it completely here, rather than deprecating it and removing it later, its a breaking change for anyone that happens to do that with today's QSVC/R. Maybe this is low risk - it would certainly need more code to deal with that. Maybe a different way along the lines to removing this might be to override repr (or str) in the interim as well to avoid using sklearns which has some problem it seems around args - I could only see repr in the sklearn source but I did not look too hard.

I see in the test it only tested against extra named arguments (kwargs) so no tests are impacted.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-machine-learning?pullRequest=220) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-machine-learning?pullRequest=220) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-machine-learning?pullRequest=220) it.</sub>

adekusar-drl commented:
@woodsp-ibm I don't fully understand your suggestion. We can deprecate them manually (I don't think `@deprecate_arguments` can handle this), e.g. we still have `*args` in `QSVC`, but if they are not empty we print a deprecation warning, ignore them and don't pass them to the super class. Is it a preferable solution here? 

I'd like to keep as less code as possible and not override `str` or `repr` from sklearn.

woodsp-ibm commented:
It maybe that the deprecation of args might need to be more manual and not use the decorator we have. Use of args can of course be detected - so we can say its deprecated and remove later. We could also call super init and include args if it was not empty otherwise call it without any args. That would allow any code that did have args to carry on working as it did - but I imagine in that case if one does a str/repr on it, it would still fail. One could in the interim override repr to return super repr but catch the error there and give back some other repr as a workaround which would be removed once the args are removed. How much is worth doing I guess depends - we only had the one error report so maybe its low risk as no-one really used args - indeed the tests we have use kwargs.

adekusar-drl commented:
I don't think we have to pass positional arguments to `SVC`, they don't use them and this leads to the issue we have right now. We can just raise a warning and discard them.

woodsp-ibm commented:
> I don't think we have to pass positional arguments to SVC, they don't use them

As far as I could tell from briefly looking at sklearn they have a deprecation decorator and end up converting any positional args to named args. Since it seems supported I do not know why the repr/str ends up failing when passed args but either way we should deprecate our exposure of args etc.

adekusar-drl commented:
@woodsp-ibm as agreed today: we deprecate `*args`, raise a deprecation warning, if arguments are not empty we pass them to the super-class and if a user prints this instance and gets an error, then they must cope with the problem on their own. I certainly don't want to override additional methods.

woodsp-ibm commented on qiskit_machine_learning/algorithms/classifiers/qsvc.py:
```diff
@@ -47,16 +48,23 @@ def __init__(self, *args, quantum_kernel: Optional[QuantumKernel] = None, **kwar
+            msg = (
+                f"Positional arguments ({args}) are deprecated as of version 0.3.0 and "
+                f"will be removed no sooner than 3 months after the release. Instead use "
+                f"keyword arguments. Positional arguments are ignored and not passed to the superclass."
```
`Positional arguments are ignored and not passed to the superclass.` This is not the case correct so should be deleted from the message

woodsp-ibm commented on qiskit_machine_learning/algorithms/regressors/qsvr.py:
```diff
@@ -45,9 +46,20 @@ def __init__(self, *args, quantum_kernel: Optional[QuantumKernel] = None, **kwar
+            msg = (
+                f"Positional arguments ({args}) are deprecated as of version 0.3.0 and "
+                f"will be removed no sooner than 3 months after the release. Instead use "
+                f"keyword arguments. Positional arguments are ignored and not passed to the superclass."
```
Similar comment for ` Positional arguments are ignored and not passed to the superclass.`

woodsp-ibm commented on releasenotes/notes/deprecated-positional-arguments-cb78dd22eb0894e0.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+issues:
+  - |
+    If positional arguments were passed into QSVR or QSVC and these classes were printed, an 	 exception was raised.
```
Presumably this is to note that if you do pass positional arguments then and use of repr/str conversion on the instance will error. This is still the case right? I.e the issues: section is to note issues that are in the release - so for me the sentence could be changed from past to present tense. If the tense was more related to this is what was fixed then it should go in the fixes: section.

woodsp-ibm commented on releasenotes/notes/deprecated-positional-arguments-cb78dd22eb0894e0.yaml:
```diff
@@ -0,0 +1,8 @@
+    If positional arguments were passed into QSVR or QSVC and these classes were printed, an 	 exception was raised.
+fixes:
+  - |
+    Positional arguments in QSVR and QSVC are deprecated.
```
This should be a `deprecations` section since it notes what is deprecated really not any fixes.

woodsp-ibm commented on test/algorithms/classifiers/test_qsvc.py:
```diff
@@ -101,6 +101,19 @@ def test_qsvc_parameters(self):
+        qsvc = QSVC(1)
+
+        with self.assertWarns(DeprecationWarning):
+            print(qsvc)
```
Can we do say `_ = str(qsvc)` in al cases so we call the string conversion that print would be doing - but unlike print it will not be printing anything to the console/log.

woodsp-ibm commented on test/algorithms/classifiers/test_qsvc.py:
```diff
@@ -101,6 +101,19 @@ def test_qsvc_parameters(self):
+        """Test QSVC deprecation warning appears when *args passed in"""
+        qsvc = QSVC(1)
+
+        with self.assertWarns(DeprecationWarning):
```
Isn't the deprecation warning issued when the class in constructed? Now it maybe that since we are passing in args that it is failing in this line like the issue reported - not sure whether we want to test it failing. For sure passing no args it should convert to str, so would print, so checking that the issue no longer exists in that case seems fine.

desireevl commented:
@woodsp-ibm, I believe I have addressed all your comments above. I am seeing though that even when no args are passed into QSVR() and QSVC(), it is unable to print the class still (so the test function fails).

woodsp-ibm commented:
I was looking at this and see a couple of hours ago scikit-learn 1.0 was released and the deprecated args that SVC had allowed seems to have been removed. 

As far as I could tell so far (with the 0.24.2) just having the args in the QSVC init signature is what ends up making it fail. I was starting to look at perhaps using a similar (or the same) deprecation decorator for the args, to see if that would work. But I do not know now given 1.0 is out - needs some further investigation I think.

adekusar-drl commented:
@woodsp-ibm please take look.
